# Clustering Neighbourhoods in Toronto

## Coursera Capstone Project

## K. Ashwini

## Part 1 - Converting Wikipedia html table into a DataFrame


In [1]:
from bs4 import BeautifulSoup as bsoup
from urllib.request import urlopen as uReq
import requests
import lxml
import pandas as pd
from pandas import DataFrame
import numpy as np

In [2]:
my_url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'


In [3]:
r=requests.get(my_url)


## Parsing the web html file with BeautifulSoup package

In [4]:
page=bsoup(r.text,"html.parser")
page


<!DOCTYPE html>

<html class="client-nojs" dir="ltr" lang="en">
<head>
<meta charset="utf-8"/>
<title>List of postal codes of Canada: M - Wikipedia</title>
<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"XpgjYgpAMNEAAtMpKBYAAAFY","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":951266227,"wgRevisionId":951266227,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Communications in Ontario","Postal codes in Canada","Toronto","Ontario-related lists"]

## 'The htlm 'page' indicates that there is a 'table class' - let's find that table


In [7]:
rtable=page.table
rtable

<table class="wikitable sortable">
<tbody><tr>
<th>Postal code
</th>
<th>Borough
</th>
<th>Neighborhood
</th></tr>
<tr>
<td>M1A
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M2A
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3A
</td>
<td>North York
</td>
<td>Parkwoods
</td></tr>
<tr>
<td>M4A
</td>
<td>North York
</td>
<td>Victoria Village
</td></tr>
<tr>
<td>M5A
</td>
<td>Downtown Toronto
</td>
<td>Regent Park / Harbourfront
</td></tr>
<tr>
<td>M6A
</td>
<td>North York
</td>
<td>Lawrence Manor / Lawrence Heights
</td></tr>
<tr>
<td>M7A
</td>
<td>Downtown Toronto
</td>
<td>Queen's Park / Ontario Provincial Government
</td></tr>
<tr>
<td>M8A
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M9A
</td>
<td>Etobicoke
</td>
<td>Islington Avenue
</td></tr>
<tr>
<td>M1B
</td>
<td>Scarborough
</td>
<td>Malvern / Rouge
</td></tr>
<tr>
<td>M2B
</td>
<td>Not assigned
</td>
<td>
</td></tr>
<tr>
<td>M3B
</td>
<td>North York
</td>
<td>Don Mills
</td></tr>
<tr>
<td>M4B
</td>
<td>Ea

In [8]:
results=rtable.find_all('tr')
nrows=len(results)
nrows

181

## Such table (named 'results' contains 181 rows. The first tow is the table header.


## Therefore, there are 181 rows of data ( index 0 to 181)


In [16]:
results[0:5]



[<tr>
 <th>Postal code
 </th>
 <th>Borough
 </th>
 <th>Neighborhood
 </th></tr>, <tr>
 <td>M1A
 </td>
 <td>Not assigned
 </td>
 <td>
 </td></tr>, <tr>
 <td>M2A
 </td>
 <td>Not assigned
 </td>
 <td>
 </td></tr>, <tr>
 <td>M3A
 </td>
 <td>North York
 </td>
 <td>Parkwoods
 </td></tr>, <tr>
 <td>M4A
 </td>
 <td>North York
 </td>
 <td>Victoria Village
 </td></tr>]

## The Header of the DataFrame is the first row of data [index 0]


In [17]:
header=results[0].text.split('\n')
header

['', 'Postal code', '', 'Borough', '', 'Neighborhood', '']

In [18]:
while("" in header) : 
    header.remove("") 

In [19]:
header

['Postal code', 'Borough', 'Neighborhood']

## Let's check some rows in order to prepare to build the loop that will extract all cells into a DataFrame. For example, let's examine row 85


In [20]:
results[85].text


'\nM4M\n\nEast Toronto\n\nStudio District\n'

In [21]:
results[85].text.split('\n')


['', 'M4M', '', 'East Toronto', '', 'Studio District', '']

In [22]:
results[85].text.split('\n')


['', 'M4M', '', 'East Toronto', '', 'Studio District', '']

In [23]:
Postcode=results[85].text.split('\n')[1]
Postcode

'M4M'

In [25]:
Borough=results[85].text.split('\n')[3]
Borough

'East Toronto'

In [26]:

Neighborhood=results[85].text.split('\n')[5]
Neighborhood

'Studio District'

## Iteration loop to extract all cells into a dataframe df


In [27]:
# iteration loop to harvest all records

records =[]
n=1
while n < nrows :
    Postcode=results[n].text.split('\n')[1]
    Borough=results[n].text.split('\n')[3]
    Neighborhood=results[n].text.split('\n')[5]
    records.append((Postcode, Borough,Neighborhood))
    n=n+1

df=pd.DataFrame(records, columns=['PostalCode', 'Borough', 'Neighbourhood'])
df.head(5)

PostalCode           Borough               Neighbourhood
0        M1A      Not assigned                            
1        M2A      Not assigned                            
2        M3A        North York                   Parkwoods
3        M4A        North York            Victoria Village
4        M5A  Downtown Toronto  Regent Park / Harbourfront

## Initial Dataframe has 180 cells or rows of data, as expected. Great!


In [28]:
df.shape


(180, 3)

In [29]:
df.tail()    #  Perfect! all data is included.


PostalCode       Borough  \
175        M5Z  Not assigned   
176        M6Z  Not assigned   
177        M7Z  Not assigned   
178        M8Z     Etobicoke   
179        M9Z  Not assigned   

                                         Neighbourhood  
175                                                     
176                                                     
177                                                     
178  Mimico NW / The Queensway West / South of Bloo...  
179

## Now, let's remove all rows with Borough = 'Not assigned' strings


### There are 77 Boroughs with 'Not assigned' strings


In [30]:
# How many rows have Borough equal to 'Not assigned'?
df[df['Borough']=='Not assigned'].count()

PostalCode       77
Borough          77
Neighbourhood    77
dtype: int64

In [31]:
# drops those rows where 'Not assigned' appears in column '[Borough]'
df1=df[~df.Borough.str.contains("Not assigned")]
df1=df1.reset_index(drop=True)

In [32]:
df1.head()

PostalCode           Borough                                 Neighbourhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [33]:
df1.tail()

PostalCode           Borough  \
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                         Neighbourhood  
98    The Kingsway / Montgomery Road  / Old Mill North  
99                                Church and Wellesley  
100              Business reply mail Processing CentrE  
101  Old Mill South / King's Mill Park / Sunnylea /...  
102  Mimico NW / The Queensway West / South of Bloo...

## The dataframe now has 103 rows after droppong all Borough with 'Not assigned'


In [39]:
df1.shape

(103, 3)

## Now, let's replace 'Not assigned' neighborhoods with the name of the Borough


In [36]:
df1.loc[df1['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = df1['Borough']


In [37]:
df1.head()           # Now, there are no rows with 'Not assigned' strings anymore


PostalCode           Borough                                 Neighbourhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [38]:
df1.shape


(103, 3)

## Now , we will consolidate the dataframe to the unique Postal Codes.


### As per instructions, we will look for repeated postalcodes and consolidate into one postalcode with the name of the neighbourhoods merged.


### Let's examine de dataframe first to set up the best loop algorithm. How many unique postalcodes , Boroughs and Neighbourhood names are there?


In [40]:
postalcodes = df1['PostalCode'].nunique()
boroughs = df1['Borough'].nunique()
neighbourhoods= df1['Neighbourhood'].nunique()
print('Unique Postalcodes : ' + str(postalcodes))
print('Unique Boroughs  : '+ str(boroughs))
print('Unique Neighbourhoods  :' + str(neighbourhoods))

Unique Postalcodes : 103
Unique Boroughs  : 10
Unique Neighbourhoods  :98


### There are 103 Unique Postcodes and 10 Unique Borough with 98 Neighbourhoods


## Let's Consolidate the dataframe to each unique PostalCodes and aggregated Neighbourhoods


In [41]:
nrows1=len(df1)
nrows1

103

## Let's keep df1 intact and let's work with df2 as precaution


In [42]:
# DataFrame df2 also has 212 rows or cells
df2=df1
df2.head()

PostalCode           Borough                                 Neighbourhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [44]:
# I will use two pointers, 'n and m' to sweep through the dataframe. So 'n' will vary from
#  0 to 102, and 'm ' will vary from 1 to 103
nrows2=len(df2)-1
nrows2

102

## Loop Algorithm to extract the consolidate dataframe df2 with unique postalcodes.


In [45]:
n=0

while n < nrows2 :
    post1=df2.iloc[n,0]
    #post1
    m=n+1
    post2=df2.iloc[m,0]
    #post2
    neigh1=df2.iloc[n,2]
    neigh2=df2.iloc[m,2]
    if post1==post2:
        df2.Neighbourhood[n,2] = neigh1=neigh1+','+neigh2
        #df2 = df2[df2.Neighbourhood != 'neigh2']
        df2=df2.drop(df2.index[m])
        nrows2=nrows2-1
        df2 = df2.reset_index(drop=True)
    else:
        n=n+1


#df2 = df2.reset_index(drop=True)
# print(post1, post2, 'Nigh1 is...'+ neigh1,',,,,,,,','Nigh2 is...' + neigh2, n,m)
df2.index

RangeIndex(start=0, stop=103, step=1)

In [46]:
df2.head()


PostalCode           Borough                                 Neighbourhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

# DataFrame consolidated to unique PostalCodes and sorted by PostalCode


In [47]:
df2.head(15)


PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   
12        M1C       Scarborough   
13        M3C        North York   
14        M4C         East York   

                                        Neighbourhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                          Regent Park / Harbourfront  
3                   Lawrence Manor / Lawrence Heights  
4        Queen's Park / Ontario Provincial Government  
5                                    Islington Avenue  
6                                     Malvern / Rouge  
7                                           Don Mills  
8                    Parkview Hill / Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park / Princess Gardens / Martin Gr...  
12           Rouge Hill / Port Union / Highland Creek  
13                                          Don Mills  
14                                   Woodbine Heights

In [48]:
df2.shape


(103, 3)

### END OF PART 1

## x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x-x


## PART 2: Now, let's get the Latitude and Longitude coordenates for every PostalCode and add those columns to the dataframe


In [52]:
import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

#import geocoder

## NOTE: I tried the geocoder but it did not work reliably, therefore I decided to use the cvs file with postalcodes provided.


In [53]:
# Read the cvs file and convert it to a dataframe

url='http://cocl.us/Geospatial_data'
df_pcodes=pd.read_csv(url)
df_pcodes.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

## Need to rename the column "Postal Code" to "PostalCode" in order to do a proper merger


In [54]:
df_pcodes.columns = ['PostalCode', 'Latitude', 'Longitude']
df_pcodes.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

## Now, let's sort our db2 dataframe by PostalCode, then we merge it with the df_pcodes created from the cvs file, based on common column of PostalCodes


In [55]:
df2s=df2.sort_values('PostalCode')
df2s.head()

PostalCode      Borough                             Neighbourhood
6         M1B  Scarborough                           Malvern / Rouge
12        M1C  Scarborough  Rouge Hill / Port Union / Highland Creek
18        M1E  Scarborough       Guildwood / Morningside / West Hill
22        M1G  Scarborough                                    Woburn
26        M1H  Scarborough                                 Cedarbrae

## Table is sorted by PostalCode


In [56]:
neighborhoods=pd.merge(df2s,df_pcodes, how='right', on = 'PostalCode')
neighborhoods.head(15)

PostalCode      Borough                                      Neighbourhood  \
0         M1B  Scarborough                                    Malvern / Rouge   
1         M1C  Scarborough           Rouge Hill / Port Union / Highland Creek   
2         M1E  Scarborough                Guildwood / Morningside / West Hill   
3         M1G  Scarborough                                             Woburn   
4         M1H  Scarborough                                          Cedarbrae   
5         M1J  Scarborough                                Scarborough Village   
6         M1K  Scarborough      Kennedy Park / Ionview / East Birchmount Park   
7         M1L  Scarborough                  Golden Mile / Clairlea / Oakridge   
8         M1M  Scarborough  Cliffside / Cliffcrest / Scarborough Village West   
9         M1N  Scarborough                       Birch Cliff / Cliffside West   
10        M1P  Scarborough  Dorset Park / Wexford Heights / Scarborough To...   
11        M1R  Scarborough                                 Wexford / Maryvale   
12        M1S  Scarborough                                          Agincourt   
13        M1T  Scarborough          Clarks Corners / Tam O'Shanter / Sullivan   
14        M1V  Scarborough  Milliken / Agincourt North / Steeles East / L'...   

     Latitude  Longitude  
0   43.806686 -79.194353  
1   43.784535 -79.160497  
2   43.763573 -79.188711  
3   43.770992 -79.216917  
4   43.773136 -79.239476  
5   43.744734 -79.239476  
6   43.727929 -79.262029  
7   43.711112 -79.284577  
8   43.716316 -79.239476  
9   43.692657 -79.264848  
10  43.757410 -79.273304  
11  43.750072 -79.295849  
12  43.794200 -79.262029  
13  43.781638 -79.304302  
14  43.815252 -79.284577

In [57]:
neighborhoods.shape


(103, 5)

## END OF PART 2

# PART 3:


## Explore and cluster the neighborhoods in Toronto


In [59]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

usage: conda-script.py [-h] [-V] command ...
conda-script.py: error: unrecognized arguments: # uncomment this line if you haven't completed the Foursquare API lab


Libraries imported.


## I will focus on a subset of all the Borough that contain the word "Toronto".


### There are 38 Borough with word Toronto.


In [60]:
toronto_data= neighborhoods[neighborhoods['Borough'].str.contains('Toronto', na = False)].reset_index(drop=True)
toronto_data.head()


PostalCode          Borough                    Neighbourhood   Latitude  \
0        M4E     East Toronto                      The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West / Riverdale  43.679557   
2        M4L     East Toronto  India Bazaar / The Beaches West  43.668999   
3        M4M     East Toronto                  Studio District  43.659526   
4        M4N  Central Toronto                    Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [61]:
toronto_data.shape


(39, 5)

## From google : the latitud and longitud for Toronto, as follows: 43.6532° N, 79.3832° W


In [62]:
latitude = 43.6532
longitude= -79.3832

## Let's create a map of Toronto with Boroughs that cointain the word Toronto


In [63]:
# create map of TORONTO using latitude and longitude values above:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

## Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.


### Define Foursquare Credentials and Version



In [64]:
CLIENT_ID = 'V4AD5UBAVYO2AOJP24XXELFUFZ0O2GXMW0CGNVHDUZOHGXHQ' # your Foursquare ID
CLIENT_SECRET = 'Q51N1BPLUIUDTQWNQGOTQLQ2XPY2QBLPBLAMCHCYNZGDS2A1' # your Foursquare Secret
VERSION = '20200416' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: V4AD5UBAVYO2AOJP24XXELFUFZ0O2GXMW0CGNVHDUZOHGXHQ
CLIENT_SECRET:Q51N1BPLUIUDTQWNQGOTQLQ2XPY2QBLPBLAMCHCYNZGDS2A1


In [65]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

In [66]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

## Now write the code to run the above function on each neighborhood and create a new dataframe called toronto_venues.


In [67]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighbourhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

The Beaches
The Danforth West / Riverdale
India Bazaar / The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park / Summerhill East
Summerhill West / Rathnelly / South Hill / Forest Hill SE / Deer Park
Rosedale
St. James Town / Cabbagetown
Church and Wellesley
Regent Park / Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond / Adelaide / King
Harbourfront East / Union Station / Toronto Islands
Toronto Dominion Centre / Design Exchange
Commerce Court / Victoria Hotel
Roselawn
Forest Hill North & West
The Annex / North Midtown / Yorkville
University of Toronto / Harbord
Kensington Market / Chinatown / Grange Park
CN Tower / King and Spadina / Railway Lands / Harbourfront West / Bathurst
Stn A PO Boxes
First Canadian Place / Underground city
Christie
Dufferin / Dovercourt Village
Little Portugal / Trinity
Brockton / Parkdale Village / Exhibition Place
High Park / The Junction South
Parkdale / Roncesva

In [68]:
print(toronto_venues.shape)
toronto_venues.head()

(1635, 7)


Neighborhood  Neighborhood Latitude  \
0                    The Beaches              43.676357   
1                    The Beaches              43.676357   
2                    The Beaches              43.676357   
3                    The Beaches              43.676357   
4  The Danforth West / Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                            Pantheon       43.677621   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351434   Greek Restaurant

## Let's check how many venues were returned for each neighborhood


In [70]:
toronto_venues.groupby('Neighborhood').count().head()


Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            56   
Brockton / Parkdale Village / Exhibition Place                         23   
Business reply mail Processing CentrE                                  16   
CN Tower / King and Spadina / Railway Lands / H...                     14   
Central Bay Street                                                     64   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             56   
Brockton / Parkdale Village / Exhibition Place                          23   
Business reply mail Processing CentrE                                   16   
CN Tower / King and Spadina / Railway Lands / H...                      14   
Central Bay Street                                                      64   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            56              56   
Brockton / Parkdale Village / Exhibition Place         23              23   
Business reply mail Processing CentrE                  16              16   
CN Tower / King and Spadina / Railway Lands / H...     14              14   
Central Bay Street                                     64              64   

                                                    Venue Longitude  \
Neighborhood                                                          
Berczy Park                                                      56   
Brockton / Parkdale Village / Exhibition Place                   23   
Business reply mail Processing CentrE                            16   
CN Tower / King and Spadina / Railway Lands / H...               14   
Central Bay Street                                               64   

                                                    Venue Category  
Neighborhood                                                        
Berczy Park                                                     56  
Brockton / Parkdale Village / Exhibition Place                  23  
Business reply mail Processing CentrE                           16  
CN Tower / King and Spadina / Railway Lands / H...              14  
Central Bay Street                                              64

## Let's find out how many unique categories can be curated from all the returned venues


In [71]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))


There are 231 uniques categories.


# Analyze Each Neighborhood


In [72]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
cols=list(toronto_onehot.columns.values)
cols.pop(cols.index('Neighborhood'))
toronto_onehot=toronto_onehot[['Neighborhood']+cols]

# rename Neighborhood for Neighbourhood so that future merge works
toronto_onehot.rename(columns = {'Neighborhood': 'Neighbourhood'}, inplace = True)
toronto_onehot.head()

Neighbourhood  Airport  Airport Food Court  Airport Gate  \
0                    The Beaches        0                   0             0   
1                    The Beaches        0                   0             0   
2                    The Beaches        0                   0             0   
3                    The Beaches        0                   0             0   
4  The Danforth West / Riverdale        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0             0         0            0                    0                 0   
1             0         0            0                    0                 0   
2             0         0            0                    0                 0   
3             0         0            0                    0                 0   
4             0         0            0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       0     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0           0              0          0         0                     0   
1           0              0          0         0                     0   
2           0              0          0         0                     0   
3           0              0          0         0                     0   
4           0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               0        0                0         0             0   
3               0        0                0         0             0   
4               0        0                0         0             0   

   Burrito Place  Bus Line  Butcher  Café  Cajun / Creole Restaurant  \
0              0         0        0     0                          0   
1              0         0        0     0                          0   
2              0         0        0     0                          0   
3              0         0        0     0                          0   
4              0         0        0     0                

In [73]:
toronto_onehot.shape


(1635, 231)

## Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category


In [74]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood   Airport  \
0                                         Berczy Park  0.000000   
1      Brockton / Parkdale Village / Exhibition Place  0.000000   
2               Business reply mail Processing CentrE  0.000000   
3   CN Tower / King and Spadina / Railway Lands / ...  0.071429   
4                                  Central Bay Street  0.000000   
5                                            Christie  0.000000   
6                                Church and Wellesley  0.000000   
7                     Commerce Court / Victoria Hotel  0.000000   
8                                          Davisville  0.000000   
9                                    Davisville North  0.000000   
10                      Dufferin / Dovercourt Village  0.000000   
11            First Canadian Place / Underground city  0.000000   
12                           Forest Hill North & West  0.000000   
13                           Garden District, Ryerson  0.000000   
14  Harbourfront East / Union Station / Toronto Is...  0.000000   
15                     High Park / The Junction South  0.000000   
16                    India Bazaar / The Beaches West  0.000000   
17        Kensington Market / Chinatown / Grange Park  0.000000   
18                                      Lawrence Park  0.000000   
19                          Little Portugal / Trinity  0.000000   
20                       Moore Park / Summerhill East  0.000000   
21                                 North Toronto West  0.000000   
22                            Parkdale / Roncesvalles  0.000000   
23       Queen's Park / Ontario Provincial Government  0.000000   
24                         Regent Park / Harbourfront  0.000000   
25                         Richmond / Adelaide / King  0.000000   
26                                           Rosedale  0.000000   
27                                           Roselawn  0.000000   
28                                Runnymede / Swansea  0.000000   
29                                     St. James Town  0.000000   
30                       St. James Town / Cabbagetown  0.000000   
31                                     Stn A PO Boxes  0.000000   
32                                    Studio District  0.000000   
33  Summerhill West / Rathnelly / South Hill / For...  0.000000   
34              The Annex / North Midtown / Yorkville  0.000000   
35                                        The Beaches  0.000000   
36                      The Danforth West / Riverdale  0.000000   
37          Toronto Dominion Centre / Design Exchange  0.000000   
38                    University of Toronto / Harbord  0.000000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0             0.000000      0.000000        0.000000         0.000000   
1             0.000000      0.000000        0.000000         0.000000   
2             0.000000      0.000000        0.000000         0.000000   
3             0.071429      0.071429        0.071429         0.142857   
4             0.000000      0.000000        0.000000         0.000000   
5             0.000000      0.000000        0.000000         0.000000   
6             0.000000      0.000000        0.000000         0.000000   
7             0.000000      0.000000        0.000000         0.000000   
8             0.000000      0.000000        0.000000         0.000000   
9             0.000000      0.000000        0.000000         0.000000   
10            0.000000      0.000000        0.000000         0.000000   
11            0.000000      0.000000        0.000000         0.000000   
12            0.000000      0.000000        0.000000         0.000000   
13            0.000000      0.000000        0.000000         0.000000   
14            0.000000      0.000000        0.000000         0.000000   
15            0.000000      0.000000        0.000000         0.000000   
16            0.000000      0.000000        0.000000         0.000000   
17            0.000000      0.000000        0

In [75]:
toronto_grouped.shape


(39, 231)

## Let's print each neighborhood along with the top 5 most common venues


In [76]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.05
1        Cocktail Bar  0.04
2  Seafood Restaurant  0.04
3      Farmers Market  0.04
4  Italian Restaurant  0.04


----Brockton / Parkdale Village / Exhibition Place----
            venue  freq
0            Café  0.13
1     Coffee Shop  0.09
2  Breakfast Spot  0.09
3       Nightclub  0.09
4    Intersection  0.04


----Business reply mail Processing CentrE----
                venue  freq
0         Yoga Studio  0.06
1             Brewery  0.06
2       Garden Center  0.06
3              Garden  0.06
4  Light Rail Station  0.06


----CN Tower / King and Spadina / Railway Lands / Harbourfront West / Bathurst----
              venue  freq
0   Airport Service  0.14
1           Airport  0.07
2      Airport Gate  0.07
3    Airport Lounge  0.07
4  Airport Terminal  0.07


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.19
1  Italian Restaurant  0.06
2      Sandwich Place  0.06
3            

## Let's put that into a pandas dataframe


## First, let's write a function to sort the venues in descending order


In [77]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

## Now let's create the new dataframe and display the top 10 venues for each neighborhood.


In [78]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighbourhood 1st Most Common Venue  \
0                                         Berczy Park           Coffee Shop   
1      Brockton / Parkdale Village / Exhibition Place                  Café   
2               Business reply mail Processing CentrE           Yoga Studio   
3   CN Tower / King and Spadina / Railway Lands / ...       Airport Service   
4                                  Central Bay Street           Coffee Shop   
5                                            Christie         Grocery Store   
6                                Church and Wellesley           Coffee Shop   
7                     Commerce Court / Victoria Hotel           Coffee Shop   
8                                          Davisville           Pizza Place   
9                                    Davisville North                   Gym   
10                      Dufferin / Dovercourt Village              Pharmacy   
11            First Canadian Place / Underground city           Coffee Shop   
12                           Forest Hill North & West                  Park   
13                           Garden District, Ryerson           Coffee Shop   
14  Harbourfront East / Union Station / Toronto Is...           Coffee Shop   
15                     High Park / The Junction South                  Café   
16                    India Bazaar / The Beaches West        Sandwich Place   
17        Kensington Market / Chinatown / Grange Park                  Café   
18                                      Lawrence Park                  Park   
19                          Little Portugal / Trinity                   Bar   
20                       Moore Park / Summerhill East            Playground   
21                                 North Toronto West        Clothing Store   
22                            Parkdale / Roncesvalles             Gift Shop   
23       Queen's Park / Ontario Provincial Government           Coffee Shop   
24                         Regent Park / Harbourfront           Coffee Shop   
25                         Richmond / Adelaide / King           Coffee Shop   
26                                           Rosedale                  Park   
27                                           Roselawn                Garden   
28                                Runnymede / Swansea                  Café   
29                                     St. James Town                  Café   
30                       St. James Town / Cabbagetown                  Park   
31                                     Stn A PO Boxes           Coffee Shop   
32                                    Studio District                  Café   
33  Summerhill West / Rathnelly / South Hill / For...           Coffee Shop   
34              The Annex / North Midtown / Yorkville                  Café   
35                                        The Beaches     Health Food Store   
36                      The Danforth West / Riverdale      Greek Restaurant   
37          Toronto Dominion Centre / Design Exchange           Coffee Shop   
38                    University of Toronto / Harbord                  Café   

   2nd Most Common Venue  3rd Most Common Venue  \
0               Beer Bar             Restaurant   
1              Nightclub            Coffee Shop   
2             Smoke Shop          Garden Center   
3                Airport        Harbor / Marina   
4     Italian Restaurant         Sandwich Place   
5                   Café                   Park   
6    Japanese Restaurant                Gay Bar   
7                   Café             Restaurant   
8           Dessert Shop         Sandwich Place   
9           Dance Studio         Breakfast Spot   
10                Bakery            Pizza Place   
11                  Café             Restaurant   
12         Jewelry Store                  Trail   
13        Clothing Store                   Café   
14              Aquarium                   Café   
15    Mexican Restaurant                    Bar   
16  Fast Foo

## Cluster Neighbourhoods


## Run k-means to cluster the neighbourhood into 5 clusters.


In [79]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

## Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.


In [80]:
toronto_merged = toronto_data

# add clustering labels
toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

PostalCode          Borough                    Neighbourhood   Latitude  \
0        M4E     East Toronto                      The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West / Riverdale  43.679557   
2        M4L     East Toronto  India Bazaar / The Beaches West  43.668999   
3        M4M     East Toronto                  Studio District  43.659526   
4        M4N  Central Toronto                    Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               0     Health Food Store                   Pub   
1 -79.352188               0      Greek Restaurant    Italian Restaurant   
2 -79.315572               0        Sandwich Place  Fast Food Restaurant   
3 -79.340923               0                  Café           Coffee Shop   
4 -79.388790               0                  Park           Swim School   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0                 Trail           Yoga Studio            Dessert Shop   
1           Coffee Shop             Bookstore  Furniture / Home Store   
2                   Gym     Food & Drink Shop           Burrito Place   
3                Bakery             Gastropub     American Restaurant   
4              Bus Line           Yoga Studio          Discount Store   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0           Event Space  Ethiopian Restaurant     Electronics Store   
1            Restaurant        Ice Cream Shop                   Spa   
2            Board Shop    Italian Restaurant            Steakhouse   
3               Brewery           Yoga Studio           Fish Market   
4    Falafel Restaurant           Event Space  Ethiopian Restaurant   

         9th Most Common Venue       10th Most Common Venue  
0  Eastern European Restaurant          Dumpling Restaurant  
1                    Juice Bar                      Brewery  
2            Fish & Chips Shop             Sushi Restaurant  
3           Italian Restaurant                    Bookstore  
4            Electronics Store  Eastern European Restaurant

# Final resulting clusters visualization - Toronto Neighborhoods


In [81]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
map_clusters

In [82]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0       East Toronto               0     Health Food Store   
1       East Toronto               0      Greek Restaurant   
2       East Toronto               0        Sandwich Place   
3       East Toronto               0                  Café   
4    Central Toronto               0                  Park   
5    Central Toronto               0                   Gym   
6    Central Toronto               0        Clothing Store   
7    Central Toronto               0           Pizza Place   
8    Central Toronto               0            Playground   
9    Central Toronto               0           Coffee Shop   
10  Downtown Toronto               0                  Park   
11  Downtown Toronto               0                  Park   
13  Downtown Toronto               0           Coffee Shop   
14  Downtown Toronto               0           Coffee Shop   
15  Downtown Toronto               0                  Café   
16  Downtown Toronto               0           Coffee Shop   
17  Downtown Toronto               0           Coffee Shop   
19  Downtown Toronto               0           Coffee Shop   
21  Downtown Toronto               0           Coffee Shop   
22   Central Toronto               0                Garden   
23   Central Toronto               0                  Park   
24   Central Toronto               0                  Café   
25  Downtown Toronto               0                  Café   
28  Downtown Toronto               0           Coffee Shop   
29  Downtown Toronto               0           Coffee Shop   
30  Downtown Toronto               0         Grocery Store   
31      West Toronto               0              Pharmacy   
32      West Toronto               0                   Bar   
33      West Toronto               0                  Café   
34      West Toronto               0                  Café   
36      West Toronto               0                  Café   
37  Downtown Toronto               0           Coffee Shop   
38      East Toronto               0           Yoga Studio   

   2nd Most Common Venue 3rd Most Common Venue    4th Most Common Venue  \
0                    Pub                 Trail              Yoga Studio   
1     Italian Restaurant           Coffee Shop                Bookstore   
2   Fast Food Restaurant                   Gym        Food & Drink Shop   
3            Coffee Shop                Bakery                Gastropub   
4            Swim School              Bus Line              Yoga Studio   
5           Dance Studio        Breakfast Spot                     Park   
6            Coffee Shop    Seafood Restaurant       Salon / Barbershop   
7           Dessert Shop        Sandwich Place         Sushi Restaurant   
8            Summer Camp                 Trail             Dessert Shop   
9                    Pub                   Spa  Health & Beauty Service   
10            Playground                 Trail             Dessert Shop   
11           Coffee Shop    Italian Restaurant                      Pub   
13                  Park                Bakery                      Pub   
14        Clothing Store                  Café          Bubble Tea Shop   
15           Coffee Shop          Cocktail Bar      American Restaurant   
16              Beer Bar            Restaurant              Cheese Shop   
17    Italian Restaurant        Sandwich Place                     Café   
19              Aquarium                  Café               Restaurant   
21                  Café            Restaurant                    Hotel   
22           Yoga Studio                 Diner           Farmers Market   
23         Jewelry Store                 Trail         Sushi Restaurant   
24        Sandwich Place           Coffee Shop             Burger Joint   
25           Yoga Studio            Restaurant                   Bakery   
28                  Café            Restaurant       Italian Restaurant   
29                  Café     

# END OF PART 3